crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+#![cfg(feature = "volta-updates")]
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::fs::{DirEntry, Metadata};
+use std::path::Path;
+
+mod empty;
+mod v0;
+mod v1;
+
+use log::debug;
+use volta_core::error::ErrorDetails;
+use volta_core::fs::read_dir_eager;
+use volta_core::layout::volta_home;
+#[cfg(unix)]
+use volta_core::layout::volta_install;
+use volta_core::shim;
+use volta_fail::{Fallible, ResultExt};
+
+enum MigrationState {
+    Empty(empty::Empty),
+    V0(Box<v0::V0>),
+    V1(Box<v1::V1>),
+}
+
+impl MigrationState {
+    fn current() -> Fallible<Self> {
+        /*
+        Triage for determining the current layout version:
+        - Does the V1 layout file exist? If yes then V1
+        - Does Volta Home exist? If yes (Windows) then V0
+            - If yes (Unix) then check if Volta Install is outside shim_dir?
+                - If yes, then V0
+                - If no, then check if $VOLTA_HOME/load.sh exists? If yes then V0
+        - Else Empty
+
+        The extra logic on Unix is necessary because Unix installs can be either inside or outside $VOLTA_HOME
+        If it is inside, then the directory necessarily must exist, so we can't use that as a determination.
+        If it is outside (and for Windows which is always outside), then if $VOLTA_HOME exists, it must be from a
+        previous, V0 installation.
+
+        Going forward, each new version will have an associated layout file, so we can use that as a discriminant
+        */
+
+        let home = volta_home()?;
```

Comment:
The `volta_home()` function will always be fallible, because determining the root involves fallible operations like determining the user's home directory from the OS (see https://github.com/volta-cli/volta/blob/master/crates/volta-core/src/layout/unix.rs#L8). Once we have the value, however, all the other operations are infallible, since they are only `Path` manipulations.